File: frontend/src/lib/ws-protocol.ts
export type WSProtocolConversionMode = "inherit" | "on" | "off"

export function normalizeWSProtocolConversion(mode?: string): WSProtocolConversionMode {
  return mode === "on" || mode === "off" ? mode : "inherit"
}

export function effectiveWSProtocolConversion(mode: string | undefined, inherited: boolean): boolean {
  const normalized = normalizeWSProtocolConversion(mode)
  return normalized === "on" || (normalized === "inherit" && inherited)
}

/** 只转换 URL 开头的 HTTP(S) scheme；已经是 WS(S) 或其他协议时保持原样。 */
export function convertHTTPToWSProtocol(url: string, enabled: boolean): string {
  if (!enabled) return url
  if (/^https:\/\//i.test(url)) return `wss://${url.slice("https://".length)}`
  if (/^http:\/\//i.test(url)) return `ws://${url.slice("http://".length)}`
  return url
}

/** 组合 WebSocket URL，并按最终生效的配置转换协议头。 */
export function wsUrl(baseUrl: string, path: string, autoConvert: boolean): string {
  const combined = !baseUrl || /^[a-z]+:\/\//i.test(path)
    ? path
    : `${baseUrl.replace(/\/$/, "")}/${path.replace(/^\//, "")}`
  return convertHTTPToWSProtocol(combined, autoConvert)
}

File: internal/models/ws_protocol.go
package models

// WSProtocolConversionMode 控制 WebSocket 接口是否自动把 HTTP(S) 协议头
// 转换为 WS(S)。除全局外的每一级都可继承上级。
type WSProtocolConversionMode string

const (
	WSProtocolConversionInherit WSProtocolConversionMode = "inherit"
	WSProtocolConversionOn      WSProtocolConversionMode = "on"
	WSProtocolConversionOff     WSProtocolConversionMode = "off"
)

// NormalizeWSProtocolConversion 将空值和未知值收敛为 inherit。
func NormalizeWSProtocolConversion(mode string) WSProtocolConversionMode {
	switch WSProtocolConversionMode(mode) {
	case WSProtocolConversionOn, WSProtocolConversionOff:
		return WSProtocolConversionMode(mode)
	default:
		return WSProtocolConversionInherit
	}
}
